packages/interactivity/src/init.js

```diff
@@ -33,10 +34,29 @@ export const initialVdom = new WeakMap();
 
 // Initialize the router with the initial DOM.
 export const init = async () => {
+	/** @type { NodeListOf<HTMLElement>} */
 	const nodes = document.querySelectorAll(
 		`[data-${ directivePrefix }-interactive]`
 	);
 
+	for ( const node of nodes ) {
+		// Before initializing the vdom, make sure a store exists for the namespace.
+		// This ensures that directives can subscribe to the store even if it has
+		// not yet been created on the client so that directives can be updated when
+		// stores are later created.
+		let namespace = /** @type {HTMLElement} */ ( node ).dataset[
+			`${ directivePrefix }Interactive`
+		];
+		try {
+			namespace = JSON.parse( namespace ).namespace;
+		} catch {}
+		if ( ! stores.has( namespace ) ) {
+			store( namespace, undefined, {
+				lock: universalUnlock,
+			} );
+		}
+	}
```

Comment:
Interesting approach. 😄

I was thinking more of checking if stores are defined when directives are evaluated. This is the place:

https://github.com/WordPress/gutenberg/blob/acd79ef3d55ca1b96bcfb908e1a10576a2797e9d/packages/interactivity/src/hooks.tsx#L260-L268

We could―before doing `stores.get( namespace )`―check if that namespace exists, and create the store at that moment if it doesn't. That would also cover the case where directives are using values with explicit namespaces that could not appear in any `data-wp-interactive`, e.g.:

```html
<div data-wp-interactive="myblock">

  <!-- namespace not in data-wp-interactive -->
  <span data-wp-text="myblock/extra::state.text"></span>

</div>
```

Reply:
> That would also cover the case where directives are using values with explicit namespaces that could not appear in any data-wp-interactive

That's interesting… I hadn't considered that there can be namespaces not associated with a data-wp-interactive directive 🤔

In that case it would be better to do this somewhere else, probably `resolve` as you suggest. I've pushed changes to revert my approach and ensure stores exist in resolve.